### ISSUE TITLE
Unhelpful error message constructing re-exported tuple struct with private fields

### ISSUE THREAD

**Issue by rumajo** [rust-lang/rust#66067]
If a public tuple struct with private fields is declared in a public module, then instantiated elsewhere, the error message is helpful. For example, the code
```rust
// src/lib.rs
pub mod a {
    pub struct Triplet(usize, usize, f64);
}
```
```rust
// src/main.rs
use triplets::a::Triplet;

fn main() {
    let mut triplets = Vec::new();
    triplets.push(Triplet(0, 0, 1.0));
}
```
causes the compiler to give the following useful error message:
```
error[E0423]: expected function, tuple struct or tuple variant, found struct `Triplet`
 --> src/main.rs:5:19
  |
5 |     triplets.push(Triplet(0, 0, 1.0));
  |                   ^^^^^^^
  |                   |
  |                   constructor is not visible here due to private fields
  |                   help: a local variable with a similar name exists: `triplets`

error: aborting due to previous error
```
This tells us exactly what the problem is - that we cannot construct `Triplet` because its fields are private.

However, if the tuple struct is in a private module, but re-exported as public, the error message isn't helpful (it's even confusing). For example,
```rust
// src/lib.rs
mod a {
    pub struct Triplet(usize, usize, f64);
}
pub use a::Triplet;
```
```rust
// src/main.rs
use triplets::Triplet;

fn main() {
    let mut triplets = Vec::new();
    triplets.push(Triplet(0, 0, 1.0));
}
```
gives the compiler error
```
error[E0423]: expected function, tuple struct or tuple variant, found struct `Triplet`
 --> src/main.rs:5:19
  |
5 |     triplets.push(Triplet(0, 0, 1.0));
  |                   ^^^^^^^
  |                   |
  |                   did you mean `Triplet { /* fields */ }`?
  |                   help: a local variable with a similar name exists: `triplets`

error: aborting due to previous error
```
which confusingly suggests initialising the tuple with braces, instead of reporting that the fields are private.

Unsurprisingly, both versions work fine if the fields of `Triplet` are made public.

**Comment by estebank** [comment]
Current output:

```
error[E0423]: expected function, tuple struct or tuple variant, found struct `Triplet`
 --> src/main.rs:5:19
  |
5 |     triplets.push(Triplet(0, 0, 1.0));
  |                   ^^^^^^^
  |                   |
  |                   constructor is not visible here due to private fields
  |                   help: a local variable with a similar name exists: `triplets`
```

**Comment by contrun** [comment]
Seems to be fixed. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=430f812f47e41c65e145054bc7e3953d

**Comment by rumajo** [comment]
This is not fixed for me on the latest nightly (2020-03-18). The bad error message doesn't seem to appear when the modules are declared and used in a single file, which is what the playground link by @contrun does.

**Comment by PitiBouchon** [comment]
Simple (reproducer)[https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=pub+use+prediction%3A%3APrediction%3B+%2F%2F+this+pub+use+is+causing+the+problem+%21%0A%0Apub+mod+prediction+%7B%0A++++%23%5Bderive%28Debug%29%5D%0A++++pub+struct+Prediction%28f32%2C+pub%28crate%29+f32%29%3B%0A++++%0A++++pub+mod+predictor+%7B%0A++++++++%2F%2F+noting%0A++++++++%0A++++++++%23%5Bcfg%28test%29%5D%0A++++++++mod+test+%7B%0A++++++++++++use+crate%3A%3APrediction%3B+%2F%2F+import+the+rexported+struct%0A++++++++++++%0A++++++++++++%23%5Btest%5D%0A++++++++++++pub+fn+my_test%28%29+%7B%0A++++++++++++++++let+x+%3D+Prediction%2842.0%2C+43.0%29%3B%0A++++++++++++++++println%21%28%22test+done%3A+%7B%3A%3F%7D%22%2C+x%29%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A%0A%0A%0Afn+main%28%29+%7B%0A++++println%21%28%22main+done%22%29+%0A%7D] 
```rust
pub use my_mod::MyStruct; // this pub use is causing the problem

mod my_mod {
    #[derive(Debug)]
    pub struct MyStruct(u32);

    mod my_sub_mod {
        use crate::MyStruct; // import the rexported struct

        fn my_func() {
            let s = MyStruct(42);
            println!("MyStruct: {:?}", s);
        }
    }
}
```
which when `cargo build` print :
```
error[E0423]: expected function, tuple struct or tuple variant, found struct `MyStruct`
  --> src/lib.rs:11:21
   |
11 |             let s = MyStruct(42);
   |                     ^^^^^^^^

For more information about this error, try `rustc --explain E0423`.
```

**Comment by n47h4n13l** [comment]
I just stumbled upon this, too.
The code PitiBouchon posted not only gives a confusing error, I would even expect it to not error at all!
Since the error refers to code in a submodule, it should be perfectly fine to use the constructor this way. For example for structs with 'normal' fields, it works perfectly.

**Comment by mtrsl** [comment]
> Simple [reproducer](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=pub+use+my_mod%3A%3AMyStruct%3B+%2F%2F+this+pub+use+is+causing+the+problem%0A%0Amod+my_mod+%7B%0A++++%23%5Bderive%28Debug%29%5D%0A++++pub+struct+MyStruct%28u32%29%3B%0A%0A++++mod+my_sub_mod+%7B%0A++++++++use+crate%3A%3AMyStruct%3B+%2F%2F+import+the+rexported+struct%0A%0A++++++++fn+my_func%28%29+%7B%0A++++++++++++let+s+%3D+MyStruct%2842%29%3B%0A++++++++++++println%21%28%22MyStruct%3A+%7B%3A%3F%7D%22%2C+s%29%3B%0A++++++++%7D%0A++++%7D%0A%7D%5D) :
> 
> pub use my_mod::MyStruct; // this pub use is causing the problem
> 
> mod my_mod {
>     #[derive(Debug)]
>     pub struct MyStruct(u32);
> 
>     mod my_sub_mod {
>         use crate::MyStruct; // import the rexported struct
> 
>         fn my_func() {
>             let s = MyStruct(42);
>             println!("MyStruct: {:?}", s);
>         }
>     }
> }
> 
> which when `cargo build` print :
> 
> ```
> error[E0423]: expected function, tuple struct or tuple variant, found struct `MyStruct`
>   --> src/lib.rs:11:21
>    |
> 11 |             let s = MyStruct(42);
>    |                     ^^^^^^^^
> 
> For more information about this error, try `rustc --explain E0423`.
> ```

I agree this is also confusing, but I'm not sure it's the same as the original bug. The original bug was about the error message not telling us about the private fields of the struct and instead suggesting the incorrect `Triplet {...}` for initialising the tuple struct.

I've just tested the original code again (the second case in the first post - the first one already had a good error message), and it actually seems to be fixed now:

```
error[E0423]: cannot initialize a tuple struct which contains private fields
 --> src/main.rs:6:19
  |
6 |     triplets.push(Triplet(0, 0, 1.0));
  |                   ^^^^^^^ help: a local variable with a similar name exists: `triplets`
  |
note: constructor is not visible here due to private fields
 --> /home/mjr/Code/test/rust-66067/src/lib.rs:3:24
  |
3 |     pub struct Triplet(usize, usize, f64);
  |                        ^^^^^  ^^^^^  ^^^ private field
  |                        |      |
  |                        |      private field
  |                        private field
```

**Comment by PitiBouchon** [comment]
I've made another issue so -> https://github.com/rust-lang/rust/issues/133343
